[5.x] Fix hyphens in JS slugs

duncanmcclean opened the pull request:
This pull request fixes an issue where hyphens weren't being handled correctly in JS slugs.

Fixes #10533.

jasonvarga commented:
This would make dashes never work. It's perfectly fine for slugs in general to have dashes. Usually that's _only_ what should be used.

What's happening in the referenced issue is that the validation wants a `handle` which doesn't allow dashes.

But we are only telling it to use `_` as a separator.

https://github.com/statamic/cms/blob/a074cb4c69cacc6f91b4063c781c5d17dd57150b/src/Fields/Field.php#L494

The JS still allows you to manually type dashes. They're valid in PHP too:

```php
Str::snake('foo bar-baz'); // foo_bar-baz
```

What we probably need is a new method that enforces changing dashes to underscores. Maybe something like:

```js
Statamic.$slug.separatedBy('_').withReplacements({'-': '_'}).create(string);
```
```yaml
replacements:
  '-': '_'
```
https://github.com/statamic/cms/blob/d140dd4e14458648b10d094a6df876e7bc8eabf7/resources/js/components/fieldtypes/SlugFieldtype.vue#L6

The `withReplacements` could pass that object into the `custom` like you've done in this PR.

duncanmcclean commented:
Good suggestion! 

I've reverted my previous fix and instead went ahead with the `replacements` approach, which I've implemented on both the PHP & JS sides.

jasonvarga commented on resources/js/components/slugs/Slug.js:
```diff
@@ -56,16 +63,20 @@ export default class Slug {
         let custom = charmap[this.#language] ?? {};
         custom["'"] = ""; // Remove apostrophes in all languages
         custom["’"] = ""; // Remove smart single quotes
-        custom[" - "] = " "; // Prevent `Block - Hero` turning into `block_-_hero`
```
Why'd this need to be removed?

duncanmcclean replied:
I think I was thinking it might not be needed after these changes...

However, on second thought, I'll add it back since not all slug/handle fields specify `replacements`. 👍

jasonvarga commented:
I realized (after you made all my requested changes 😬) that in PHP's `Str::slug()` if you were to use `_` as a separator, `JSON-LD` would be converted to `json-ld` automatically.

So, I reverted the `withReplacements` stuff and instead just make dashes convert to the separator, when the separator isn't a dash. It matches PHP.

Everything you did was great and worked well, but I don't think we need it.
